### PR TITLE
In use cleanup

### PR DESCRIPTION
**Pull request by zsinx6**
#### What
Implement in use cleanup.

#### Why
We need to clean the in use licenses from the booking table.

`Task`: https://omnivector-solutions.monday.com/boards/1204470680/pulses/1456713032

### CODE REVIEW

File: agent/lm_agent/main.py [hunk 1]
```diff
@@ -18,7 +21,7 @@
 AGENT_VERSION = pkg_resources.get_distribution("license-manager-agent").version
 
 
-app = FastAPI()
+app: Any = FastAPI()
```
Comment (dusktreader) [inline agent/lm_agent/main.py]: Shouldn't the type for `app` be `FastAPI`?

File: agent/lm_agent/main.py [hunk 2]
```diff
@@ -103,3 +106,10 @@ async def collect_stats():
 
 
 app.include_router(api_v1, prefix="/api/v1")
+
+if settings.SENTRY_DSN:
+    sentry_sdk.init(
+        dsn=settings.SENTRY_DSN,
+        traces_sample_rate=1.0,
+    )
+    app = SentryAsgiMiddleware(app)
```
Comment (dusktreader) [inline agent/lm_agent/main.py]: I think the idiomatic FastAPI way to do this is:

```python
app.add_middleware(SentryAsgiMiddleware)
```

see: https://fastapi.tiangolo.com/advanced/middleware/#adding-asgi-middlewares

File: agent/lm_agent/reconciliation.py [hunk 3]
```diff
@@ -16,18 +16,14 @@
...
-    """Unable to delete booking row."""
-
-
 async def remove_booked_for_job_id(job_id: str):
     """
     Send DELETE to /api/v1/booking/book/{job_id}.
     """
-    response = await async_client().delete("/api/v1/booking/book{job_id}")
+    response = await async_client().delete(f"/api/v1/booking/book/{job_id}")
+    logger.debug(f"response from delete: {response.__dict__}")
```
Comment (dusktreader) [inline agent/lm_agent/reconciliation.py]: do we want to leave this in, or was this only for debugging?

File: agent/lm_agent/reconciliation.py [hunk 4]
```diff
@@ -53,38 +49,51 @@ def get_greatest_grace_time(job_id: str, grace_times: Dict[int, int], booking_ro
     """
     Find the greatest grace_time for the given job_id.
     """
-    greatest_grace_time = 0
+    greatest_grace_time = -1
```
Comment (dusktreader) [inline agent/lm_agent/reconciliation.py:51]: Should an exception be raised if, after traversing the booking rows, greatest_grase_time is still -1?

File: agent/lm_agent/reconciliation.py [hunk 5]
```diff
@@ -53,38 +49,51 @@ def get_greatest_grace_time(job_id: str, grace_times: Dict[int, int], booking_ro
...
+        if not book:
             continue
-        config_id_for_grace_times = book["config_id"]
-        if grace_times[config_id_for_grace_times] > greatest_grace_time:
-            greatest_grace_time = grace_times[config_id_for_grace_times]
+        for inner_book in book:
+            if str(inner_book["job_id"]) != str(job_id):
+                continue
+            config_id_for_grace_times = inner_book["config_id"]
+            if grace_times[config_id_for_grace_times] > greatest_grace_time:
```
Comment (dusktreader) [inline agent/lm_agent/reconciliation.py]: you could replace these two lines with:
```python
greatest_grace_time = max(greatest_gracetime, grace_times[config_id_for_grace_times])
```
I think it might read nicely this way

File: agent/lm_agent/reconciliation.py [hunk 6]
```diff
@@ -53,38 +49,51 @@ def get_greatest_grace_time(job_id: str, grace_times: Dict[int, int], booking_ro
...
+                greatest_grace_time = grace_times[config_id_for_grace_times]
     return greatest_grace_time
 
 
+def get_running_jobs(squeue_result: List) -> List:
+    squeue_running_jobs = []
+    for job in squeue_result:
```
Comment (dusktreader) [inline agent/lm_agent/reconciliation.py]: I think this logic could be reduced with a list comprehension (or filter() ):
```python
return [j for j in squeue_result if j["state"] == "RUNNING"]
```

File: agent/lm_agent/reconciliation.py [hunk 7]
```diff
@@ -53,38 +49,51 @@ def get_greatest_grace_time(job_id: str, grace_times: Dict[int, int], booking_ro
...
+    squeue_running_jobs = get_running_jobs(squeue_result)
+
     grace_times = await get_all_grace_times()
     get_booked_call = []
     for job in squeue_running_jobs:
         job_id = job["job_id"]
         get_booked_call.append(get_booked_for_job_id(job_id))
 
-    booking_rows_for_running_jobs = asyncio.gather(*get_booked_call)
+    booking_rows_for_running_jobs = await asyncio.gather(*get_booked_call, return_exceptions=True)
```
Comment (dusktreader) [inline agent/lm_agent/reconciliation.py]: How does the loop below handle gathered exception results?

File: agent/lm_agent/workload_managers/slurm/cmd_utils.py [hunk 8]
```diff
@@ -60,25 +60,39 @@ class LicenseBookingRequest(BaseModel):
     lead_host: str
 
 
-async def get_required_licenses_for_job(slurm_job_id: str) -> List:
+async def get_required_licenses_for_job(slurm_job_id: str, user_name: str, lead_host: str) -> Any:
```
Comment (dusktreader) [inline agent/lm_agent/workload_managers/slurm/cmd_utils.py]: Seems like the return type should be LicenseBookingRequest

File: agent/lm_agent/workload_managers/slurm/cmd_utils.py [hunk 9]
```diff
@@ -60,25 +60,39 @@ class LicenseBookingRequest(BaseModel):
...
 
-    required_features = list()
+    license_booking_request = LicenseBookingRequest(
+        job_id=slurm_job_id,
+        bookings=[],
+        user_name=user_name,
+        lead_host=lead_host,
+    )
+    if not license_array:
+        return []
```
Comment (dusktreader) [inline agent/lm_agent/workload_managers/slurm/cmd_utils.py]: I don't understand why we would return an empty list here. Should it be None instead since a successful lookup returns a single instance of LicenseBookingRequest?

File: agent/lm_agent/workload_managers/slurm/cmd_utils.py [hunk 10]
```diff
@@ -60,25 +60,39 @@ class LicenseBookingRequest(BaseModel):
...
 
     if license_array[0] != "(null)":
         for requested_license in license_array:
             license_regex = re.compile(r"(\w+)\.(\w+)@(\w+):(\d+)")
             if license_regex.match(requested_license):
                 # If the regex matches, parse the values
-                product_feature, _ = requested_license.split("@")
+                product_feature, license_server_tokens = requested_license.split("@")
```
Comment (dusktreader) [inline agent/lm_agent/workload_managers/slurm/cmd_utils.py]: Seems like we should be using the match groups for this instead of using `split()`.

Also, I would like to see this logic split out into a function with a set of unit tests that bash on the edge-cases a little bit

File: backend/lm_backend/api/license.py [hunk 11]
```diff
@@ -83,12 +90,57 @@ async def _get_these_licenses(product_features: Sequence[str]) -> List[LicenseUs
...
 
+async def delete_if_in_use_booking(license: LicenseUseReconcileRequest):
+    """
+    Check the database for the license, if it is booked, then delete it.
+    """
+    queries = []
+    for used_license in license.used_licenses:
+        query = (
+            booking_table.select()
+            .where(booking_table.c.lead_host == used_license["lead_host"])
```
Comment (dusktreader) [inline backend/lm_backend/api/license.py:101]: Nitpicky comment: you can send all three expressions to a single where:

```python
booking_table.select().where(
    booking_table.c.lead_host == used_kicense["lead_host"],
    booking_table.c.user_name == used_kicense["user_name"],
    booking_table.c.booked == used_kicense["booked"],
)
```

File: backend/lm_backend/api/license.py [hunk 12]
```diff
@@ -83,12 +90,57 @@ async def _get_these_licenses(product_features: Sequence[str]) -> List[LicenseUs
...
+            .where(booking_table.c.user_name == used_license["user_name"])
+            .where(booking_table.c.booked == used_license["booked"])
+        )
+        queries.append(database.fetch_one(query))
+    fetched = await asyncio.gather(*queries)
+    bookings = []
+    for item in fetched:
```
Comment (dusktreader) [inline backend/lm_backend/api/license.py]: maybe collapse into list comprehension

**Review by dusktreader — CHANGES_REQUESTED** [review]

File: agent/lm_agent/reconciliation.py (hunk 3 above)
Comment (zsinx6) [reply agent/lm_agent/reconciliation.py]: To leave it in

Comment (jamesbeedy) [reply agent/lm_agent/reconciliation.py]: > do we want to leave this in, or was this only for debugging?

Both :)

File: agent/lm_agent/reconciliation.py (hunk 4 above)
Comment (zsinx6) [reply agent/lm_agent/reconciliation.py:51]: I'm not sure which way is better for readability, don't know if handling the exception in the outside function is good as well

File: agent/lm_agent/reconciliation.py (hunk 7 above)
Comment (zsinx6) [reply agent/lm_agent/reconciliation.py]: It doesn't, will remove this flag, if any exception happens in this method, something is wrong and should to go sentry

File: backend/lm_backend/api/license.py (hunk 11 above)
Comment (zsinx6) [reply backend/lm_backend/api/license.py:101]: Actually not really:
`TypeError: where() takes 2 positional arguments but 5 were given`, and also I think this is more easy to add things if easy "operation" (e.g. where) is in its own line

File: agent/lm_agent/reconciliation.py (hunk 4 above)
Comment (dusktreader) [reply agent/lm_agent/reconciliation.py:51]: Fair enough. I think usage will illuminate if we need an exception or not. This is good for now

File: backend/lm_backend/api/license.py (hunk 11 above)
Comment (dusktreader) [reply backend/lm_backend/api/license.py:101]: Welp. The databases docs say they support sqlalchemy API, but it doesn't seem to line up. Good the way it was!

File: backend/tests/api/test_license.py [hunk 13]
```diff
@@ -141,3 +144,114 @@ async def test_map_bookings(some_licenses, insert_objects):
...
+        {"booked": 13, "lead_host": "host1", "user_name": "user1"},
+        {"booked": 14, "lead_host": "host1", "user_name": "user1"},
+    ]
+    license_reconcile_request = LicenseUseReconcileRequest(
+        used=19, product_feature="hello.world", total=100, used_licenses=used_licenses
+    )
+
+    await license._delete_if_in_use_booking(license_reconcile_request)
+
+    booking_rows = [BookingRow.parse_obj(x) for x in await database.fetch_all(booking_table.select())]
```
Comment (dusktreader) [inline backend/tests/api/test_license.py]: I don't think we need to parse the response in this test. We could just count the results

File: backend/tests/api/test_license.py [hunk 14]
```diff
@@ -141,3 +144,114 @@ async def test_map_bookings(some_licenses, insert_objects):
...
+    await insert_objects(some_licenses, license_table)
+
+    used_licenses = []
+    license_reconcile_request = LicenseUseReconcileRequest(
+        used=19, product_feature="hello.world", total=100, used_licenses=used_licenses
+    )
+
+    await license._delete_if_in_use_booking(license_reconcile_request)
+
+    booking_rows = [BookingRow.parse_obj(x) for x in await database.fetch_all(booking_table.select())]
```
Comment (dusktreader) [inline backend/tests/api/test_license.py]: Same here. No need to parse

File: backend/tests/api/test_license.py [hunk 15]
```diff
@@ -141,3 +144,114 @@ async def test_map_bookings(some_licenses, insert_objects):
...
+    license_reconcile_request = LicenseUseReconcileRequest(
+        used=19, product_feature="hello.world", total=100, used_licenses=used_licenses
+    )
+
+    response = await backend_client.patch(
+        "/api/v1/license/reconcile", json=[license_reconcile_request.dict()]
+    )
+    assert response.status_code == status.HTTP_200_OK
+
+    booking_rows = [BookingRow.parse_obj(x) for x in await database.fetch_all(booking_table.select())]
```
Comment (dusktreader) [inline backend/tests/api/test_license.py]: extra parse()

**Review by dusktreader — CHANGES_REQUESTED** [review]
A few more comments for you

File: agent/lm_agent/workload_managers/slurm/slurmctld_epilog.py [hunk 16]
```diff
@@ -42,8 +42,10 @@ async def main():
     # Acqure the job context and get the job_id.
     ctxt = get_job_context()
     job_id = ctxt["job_id"]
+    user_name = ctxt["user_name"]
+    lead_host = ctxt["lead_host"]
 
-    license_booking_request = await get_required_licenses_for_job(job_id)
+    license_booking_request = await get_required_licenses_for_job(job_id, user_name, lead_host)
```
Comment (jamesbeedy) [inline agent/lm_agent/workload_managers/slurm/slurmctld_epilog.py]: Since we don't use `user_name` or `grace_time` in the `get_required_licenses_for_job()` function, possibly it is better to have it return a list of `LicenseBooking` instead of a `LicenseBookingRequest`?

File: backend/lm_backend/api/license.py [hunk 17]
```diff
@@ -83,12 +90,54 @@ async def _get_these_licenses(product_features: Sequence[str]) -> List[LicenseUs
...
+    Check the database for the license, if it is booked, then delete it.
+    """
+    queries = []
+    for used_license in license.used_licenses:
+        query = (
+            booking_table.select()
+            .where(booking_table.c.lead_host == used_license["lead_host"])
+            .where(booking_table.c.user_name == used_license["user_name"])
+            .where(booking_table.c.booked == used_license["booked"])
+            .where(booking_table.c.product_feature == license.product_feature)
```
Comment (jamesbeedy) [inline backend/lm_backend/api/license.py:104]: 👍

**Review by jamesbeedy — CHANGES_REQUESTED** [review]
Looking good! Resolve the comments and then I'll be +1.

File: agent/lm_agent/reconciliation.py [hunk 18]
```diff
@@ -16,18 +16,14 @@
...
-class FailedToRemoveBookedViaGraceTime(Exception):
-    """Unable to delete booking row."""
-
-
 async def remove_booked_for_job_id(job_id: str):
     """
     Send DELETE to /api/v1/booking/book/{job_id}.
     """
-    response = await async_client().delete("/api/v1/booking/book{job_id}")
+    response = await async_client().delete(f"/api/v1/booking/book/{job_id}")
```
Comment (matheushent) [inline agent/lm_agent/reconciliation.py:23]: shouldn't this be a redirection instead of a direct call?

File: agent/lm_agent/reconciliation.py [hunk 19]
```diff
@@ -44,47 +40,54 @@ async def get_booked_for_job_id(job_id: str) -> Dict:
     """
     Return the booking row for the given job_id.
     """
-    response = await async_client().get("/api/v1/booking/job/{job_id}")
-    book_row = response.json()
-    return book_row
+    response = await async_client().get(f"/api/v1/booking/job/{job_id}")
```
Comment (matheushent) [inline agent/lm_agent/reconciliation.py:43]: same comment about redirection

File: agent/lm_agent/reconciliation.py (hunk 18 above)
Comment (zsinx6) [reply agent/lm_agent/reconciliation.py:23]: Can you explain more?

File: agent/lm_agent/workload_managers/slurm/cmd_utils.py [hunk 20]
```diff
@@ -92,9 +108,11 @@ async def check_feature_token_availablity(lbr: LicenseBookingRequest) -> bool:
     with httpx.Client() as client:
         resp = client.get(f"{settings.AGENT_BASE_URL}/api/v1/license/all", headers=LM2_AGENT_HEADERS)
```
Comment (matheushent) [inline agent/lm_agent/workload_managers/slurm/cmd_utils.py:119]: same comment about redirection

File: agent/lm_agent/reconciliation.py (hunk 18 above)
Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: I'd like to see `RedirectResponse` from starlette being used here. It seems more coherent to me. Thoughts?

Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: It seems to me that doing the current way the API iteself increases its traffic. It seems inefficient some way

File: agent/lm_agent/workload_managers/slurm/slurmctld_epilog.py (hunk 16 above)
Comment (zsinx6) [reply agent/lm_agent/workload_managers/slurm/slurmctld_epilog.py]: Actually we build the entire license_booking_request using the `user_name` and `lead_host`. If we build the object outside the function it will only move the logic around, I don't see the point of modify this

File: agent/lm_agent/reconciliation.py (hunk 18 above)
Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: > Can you explain more?

The current way the API calls itself, which I think may be inefficient since the API would be increasing its own traffic. Perhaps we could improve it using `RedirectResponse` from starlette

Comment (zsinx6) [reply agent/lm_agent/reconciliation.py:23]: But why redirect? We need to actually make a call to that endpoint

Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: yep, but internally it doesn't necessary mean a call but a redirection. I wonder the efficiency of this. We can improve that

Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: redirection instead of direct call means less network traffic

Comment (zsinx6) [reply agent/lm_agent/reconciliation.py:23]: No, the api will not call itself, it is going to call the backend

Comment (zsinx6) [reply agent/lm_agent/reconciliation.py:23]: Take a look at the async_client: https://github.com/omnivector-solutions/license-manager/blob/75829ed873bc0f3c1c66902da5a8ba08e7ccc326/agent/lm_agent/forward.py#L25

Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: I still think this can be improved to a internal redirection instead of a local network call. But anyway, I'd like more thoughts on that

Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: I couldn't see in starlette's code whether this is done the same way

Comment (zsinx6) [reply agent/lm_agent/reconciliation.py:23]: The backend is in AWS, it must be a outside call

Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: Perhaps I meant confusedly. I meant not returning the redirect response but using the RedirectResponse just to call the API internally

Comment (matheushent) [reply agent/lm_agent/reconciliation.py:23]: oh now I see, it doesn't make sense what I'm saying. Just ignore it

**Review by dusktreader — APPROVED** [review]
Nice work!

File: agent/lm_agent/workload_managers/slurm/slurmctld_epilog.py (hunk 16 above)
Comment (jamesbeedy) [reply agent/lm_agent/workload_managers/slurm/slurmctld_epilog.py]: I guess I wasn't very clear ... my thoughts are basically that if we don't need the `LicenseBookingRequest` object, we just need the bookings, then why pass this information around and create this `LicenseBookingRequest` when we can just change the return type of the function and have it only return a list of bookings?

**Review by jamesbeedy — APPROVED** [review]
Thank you for resolving this! 👍
